flocker/control/_protocol.py

```diff
@@ -663,35 +725,105 @@ def _send_state_to_connections(self, connections):
+        Send the latest cluster configuration and state to ``connection``.
 
         :param ControlAMP connection: The connection to use to send the
             command.
-
-        :param Deployment configuration: The cluster configuration to send.
-        :param DeploymentState state: The current cluster state to send.
         """
+
+        # Set the configuration and the state to the latest versions. It is
+        # okay to call this even if the latest configuration is the same
+        # object.
+        self._configuration_generation_tracker.insert_latest(configuration)
+        self._state_generation_tracker.insert_latest(state)
+
         action = LOG_SEND_TO_AGENT(agent=connection)
         with action.context():
-            # Use ``maybeDeferred`` so if an exception happens,
-            # it will be wrapped in a ``Failure`` - see FLOC-3221
-            d = DeferredContext(maybeDeferred(
-                connection.callRemote,
-                ClusterStatusCommand,
-                configuration=configuration,
-                state=state,
-                eliot_context=action
-            ))
-            d.addActionFinish()
+
+            # Attempt to compute a diff to send to the connection
+            last_received_generations = (
+                self._last_recieved_generation[connection]
+            )
+
+            config_gen_tracker = self._configuration_generation_tracker
+            configuration_diff = (
+                config_gen_tracker.get_diff_from_hash_to_latest(
+                    last_received_generations.config_hash
+                )
+            )
+
+            state_gen_tracker = self._state_generation_tracker
+            state_diff = (
+                state_gen_tracker.get_diff_from_hash_to_latest(
+                    last_received_generations.state_hash
+                )
+            )
+
+            if configuration_diff is not None and state_diff is not None:
+                # If both diffs were successfully computed, send a command to
+                # send the diffs along with before and after hashes so the
+                # nodes can verify the application of the diffs.
+                d = DeferredContext(maybeDeferred(
+                    connection.callRemote,
+                    ClusterStatusDiffCommand,
+                    configuration_diff=configuration_diff,
+                    start_configuration_generation=(
+                        last_received_generations.config_hash
+                    ),
+                    end_configuration_generation=(
+                        config_gen_tracker.get_latest_hash()
+                    ),
+                    state_diff=state_diff,
+                    start_state_generation=(
+                        last_received_generations.state_hash
+                    ),
+                    end_state_generation=state_gen_tracker.get_latest_hash(),
+                    eliot_context=action
+                ))
+                d.addActionFinish()
```

Comment:
:question: I guess we could start logging the configuration again since they'll be much smaller and less frequent now.